### PR TITLE
Add BigQuery service implementation

### PR DESCRIPTION
**Pull request by mziccard**
This class implements the `BigQuery` service, main changes are:
- Connect `DefaultBigQueryRpc` and `BigQueryImpl` to `BigQueryOptions`
- Add `defaultInstance` method to `BigQueryOptions`
- Add `Option` base class for `BigQuery` operation options and test
- Add methods and options to the `BigQuery` interface
- Add `BigQueryImpl` service implementation
- Add `BigQueryImplTest` with unit tests

### CODE REVIEW

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 1]
```diff
@@ -25,5 +34,602 @@
  */
 public interface BigQuery extends Service<BigQueryOptions> {
 
-  // TODO(mziccard) add missing methods
+  enum DatasetField {
+    ACCESS("access"),
+    CREATION_TIME("creationTime"),
+    DATASET_REFERENCE("datasetReference"),
+    DEFAULT_TABLE_EXPIRATION_MS("defaultTableLifetime"),
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: defaultTableExpirationMs ?

Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Where did I get that from?:D Will fix!

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 2]
```diff
@@ -25,5 +34,602 @@
...
+    DatasetField(String selector) {
+      this.selector = selector;
+    }
+
+    public String selector() {
+      return selector;
+    }
+
+    static String selector(DatasetField... fields) {
+      HashSet<String> fieldStrings = Sets.newHashSetWithExpectedSize(fields.length + 1);
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Just thinking out loud. If enum fields implemented a "Field" interface with a `String selector()` method then you can have only one static selector method `static String selector(Selector first, Selector... other)` for all...

Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Yes and we should probably reuse that same interface in `Storage` and other services that use field selection. Do you mind if we make this part of a separate PR once bigquery is merged? That would make easier to put the interface in `gcloud-java-core` and modify also `gcloud-java-storage` (no conflicts on merge)

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 3]
```diff
@@ -25,5 +34,602 @@
...
+      fieldStrings.add(DATASET_REFERENCE.selector());
+      for (DatasetField field : fields) {
+        fieldStrings.add(field.selector());
+      }
+      return com.google.common.base.Joiner.on(',').join(fieldStrings);
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Any reason for this fully qualified reference (and avoiding the import)?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 2 above)
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Nit, `Set<String> fieldStrings`

Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Separate PR sounds fine to me.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 3 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: No actual reason

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 4]
```diff
@@ -25,5 +34,602 @@
  */
 public interface BigQuery extends Service<BigQueryOptions> {
 
-  // TODO(mziccard) add missing methods
+  enum DatasetField {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Should we have javadoc for the enums `DatasetField`, `TableField`, etc?

Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: We don't have it for storage (but this is no good reason for not adding it). I think that the names are quite self-explanatory. I also think that adding javadoc for each enum value would mean adding too much "obvious" comments. I'm +1 to adding a javadoc comment on top of the enum type with a link to a dataset/table/job resource definition.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 5]
```diff
@@ -25,5 +34,602 @@
...
+
+    private static final long serialVersionUID = -7166083569900951337L;
+
+    private DatasetDeleteOption(BigQueryRpc.Option option, Object value) {
+      super(option, value);
+    }
+
+    /**
+     * Returns an option to delete a dataset even if non-empty. If not provided, attempting to
+     * delete a non-empty dataset will result in a {@link BigQueryException} being thrown.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Is there a way to tell from the exception that the delete failed for that reason?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 4 above)
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: I wasn't bothered by it because I also think that name level comments are redundant. I didn't mind much about the class level as their usage is always part of a specific `Option` and they have a nice javadoc for the `fields` method however it would be nice to have a class level javadoc for consistency and for lulling checkstyle and a like about it.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 6]
```diff
@@ -25,5 +34,602 @@
...
+   * {@link #getJob(String, JobOption...)}) to see if the cancel operation completed successfully.
+   *
+   * @return {@code true} if cancel was requested successfully, {@code false} if the job was not
+   *     found
+   * @throws BigQueryException upon failure
+   */
+  boolean cancel(JobId tableId) throws BigQueryException;
+
+  /**
+   * Runs the query associated to the request.
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Nit: replace "to" with "with"

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 7]
```diff
@@ -25,5 +34,602 @@
...
+   */
+  boolean delete(DatasetId datasetId, DatasetDeleteOption... options) throws BigQueryException;
+
+  /**
+   * Deletes the requested table.
+   *
+   * @return {@code true} if table was deleted, {@code false} if it was not found.
+   * @throws BigQueryException upon failure
+   */
+  boolean delete(String datasetId, String tableId) throws BigQueryException;
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: There is no support for conditional delete based on etag?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 8]
```diff
@@ -25,5 +34,602 @@
...
+  /**
+   * Lists the dataset's tables.
+   *
+   * @throws BigQueryException upon failure
+   */
+  Page<JobInfo> listJobs(JobListOption... options) throws BigQueryException;
+
+  /**
+   * Sends a job cancel request. This call will return immediately, and the client will need to poll
+   * for the job status to see if the cancel completed successfully.
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Nit: we could add the links like in the comment below

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 9]
```diff
@@ -25,5 +34,602 @@
...
+    public static TableDataListOption startPageToken(String pageToken) {
+      return new TableDataListOption(BigQueryRpc.Option.PAGE_TOKEN, pageToken);
+    }
+
+    /**
+     * Returns an option that sets the zero-based index of the row from which to start listing table
+     * data.
+     */
+    public static TableDataListOption startIndex(long index) {
+      return new TableDataListOption(BigQueryRpc.Option.START_INDEX, index);
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Should we `checkArgument(index >= 0)`? If so, we could also apply it for `maxResults > 0` and other similar places.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 10]
```diff
@@ -25,5 +34,602 @@
...
+
+    /**
+     * Returns an option to list only jobs that match the provided filters.
+     */
+    public static JobListOption stateFilter(JobStatus.State... stateFilters) {
+      List<String> stringFilters = Lists.transform(ImmutableList.copyOf(stateFilters),
+          new Function<JobStatus.State, String>() {
+            @Override
+            public String apply(JobStatus.State state) {
+              return state.toString().toLowerCase();
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: nit, `name()` is slightly better than `toString()` as the latter could be overridden.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 11]
```diff
@@ -25,5 +34,602 @@
...
+
+    /**
+     * Returns an option to specify the job's fields to be returned by the RPC call. If this option
+     * is not provided all job's fields are returned. {@code JobOption.fields()}) can be used to
+     * specify only the fields of interest. {@link JobInfo#jobId()}, {@link JobStatus#state()},
+     * {@link JobStatus#error()} as well as type-specific configuration (e.g.
+     * {@link QueryJobInfo#query()} for Query Jobs) are always returned, even if not specified.
+     * {@link JobField#SELF_LINK} and {@link JobField#ETAG} can not be selected when listing jobs.
+     */
+    public static JobListOption fields(JobField... fields) {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Do you think this is useful (considering the possible return values and the minimum that we specify)? I am not sure about it, and not having it would be consistent with the other listing operations that do not provide a field selector.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 12]
```diff
@@ -25,5 +34,602 @@
...
+
+    /**
+     * Returns an option to specify the maximum number of rows to be returned.
+     */
+    public static QueryResultsOption maxResults(long maxResults) {
+      return new QueryResultsOption(BigQueryRpc.Option.MAX_RESULTS, maxResults);
+    }
+
+    /**
+     * Returns an option to specify the page token from which to start listing query results.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: is "listing" the right term here?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 13]
```diff
@@ -25,5 +34,602 @@
...
+     */
+    public static QueryResultsOption startPageToken(String pageToken) {
+      return new QueryResultsOption(BigQueryRpc.Option.PAGE_TOKEN, pageToken);
+    }
+
+    /**
+     * Returns an option that sets the zero-based index of the row from which to start listing query
+     * results.
+     */
+    public static QueryResultsOption startIndex(Long startIndex) {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Why using `Long` (here and below).

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 14]
```diff
@@ -25,5 +34,602 @@
...
+
+    private static final long serialVersionUID = 1674133909259913250L;
+
+    private DatasetOption(BigQueryRpc.Option option, Object value) {
+      super(option, value);
+    }
+
+    /**
+     * Returns an option to specify the dataset's fields to be returned by the RPC call. If this
+     * option is not provided all dataset's fields are returned. {@code DatasetOption.fields}) can
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Nit: remove ')' after `fields}`

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 15]
```diff
@@ -25,5 +34,602 @@
...
+
+    private static final long serialVersionUID = -1723870134095936772L;
+
+    private TableOption(BigQueryRpc.Option option, Object value) {
+      super(option, value);
+    }
+
+    /**
+     * Returns an option to specify the table's fields to be returned by the RPC call. If this
+     * option is not provided all table's fields are returned. {@code TableOption.fields}) can be
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Nit: there's an extra ')' after '.fields}'

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 16]
```diff
@@ -25,5 +34,602 @@
...
+
+    /**
+     * Returns an option to list all jobs, even the ones issued by other users.
+     */
+    public static JobListOption allUsers() {
+      return new JobListOption(BigQueryRpc.Option.ALL_USERS, true);
+    }
+
+    /**
+     * Returns an option to list only jobs that match the provided filters.
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Should we specify that the filter is on the job's state?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 17]
```diff
@@ -25,5 +34,602 @@
...
+    /**
+     * Returns an option to specify the page token from which to start listing jobs.
+     */
+    public static JobListOption startPageToken(String pageToken) {
+      return new JobListOption(BigQueryRpc.Option.PAGE_TOKEN, pageToken);
+    }
+
+    /**
+     * Returns an option to specify the job's fields to be returned by the RPC call. If this option
+     * is not provided all job's fields are returned. {@code JobOption.fields()}) can be used to
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: extra `)` after @code block

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 18]
```diff
@@ -25,5 +34,602 @@
...
+
+    private static final long serialVersionUID = -3111736712316353665L;
+
+    private JobOption(BigQueryRpc.Option option, Object value) {
+      super(option, value);
+    }
+
+    /**
+     * Returns an option to specify the job's fields to be returned by the RPC call. If this option
+     * is not provided all job's fields are returned. {@code JobOption.fields}) can be used to
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: extra ')'

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 19]
```diff
@@ -25,5 +34,602 @@
...
+  /**
+   * Returns the requested dataset or {@code null} if not found.
+   *
+   * @throws BigQueryException upon failure
+   */
+  DatasetInfo getDataset(DatasetId datasetId, DatasetOption... options) throws BigQueryException;
+
+  /**
+   * Lists the project's datasets. This method returns partial information on each dataset
+   * ({@link DatasetInfo#datasetId()} ()}, {@link DatasetInfo#friendlyName()} and
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Extra `()}` after `datasetId`

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Option.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+import com.google.common.base.MoreObjects;
+import com.google.gcloud.spi.BigQueryRpc;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Base class for BigQuery operation option.
+ */
+class Option implements Serializable {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/Option.java]: Should we add `Option`'s subclasses to serialization tests?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/OptionTest.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+
+  @Test
+  public void testOption() {
+    Option option = new Option(BigQueryRpc.Option.PAGE_TOKEN, "token");
+    assertEquals(BigQueryRpc.Option.PAGE_TOKEN, option.rpcOption());
+    assertEquals("token", option.value());
+  }
+
+  @Test(expected=NullPointerException.class)
+  public void testIndexOutOfBoundsException() {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/OptionTest.java]: testIndexOutOfBoundsException -> testNullOption?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 22]
```diff
@@ -0,0 +1,982 @@
...
+  private BigQueryOptions options;
+  private BigQueryRpcFactory rpcFactoryMock;
+  private BigQueryRpc bigqueryRpcMock;
+  private BigQuery bigquery;
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Before
+  public void setUp() throws IOException, InterruptedException {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Nit: we can probably get rid of `throws ...`

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 23]
```diff
@@ -0,0 +1,982 @@
...
+      .build();
+
+  // Empty BigQueryRpc options
+  private static final Map<BigQueryRpc.Option, ?> EMPTY_RPC_OPTIONS = ImmutableMap.of();
+
+  // Dataset options
+  private static final BigQuery.DatasetOption DATASET_OPTION_FIELDS =
+      BigQuery.DatasetOption.fields(BigQuery.DatasetField.ACCESS, BigQuery.DatasetField.ETAG);
+
+  // Dataset list options
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: We could add the "max results" option here and set it to something low for testing purposes

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java [hunk 24]
```diff
@@ -0,0 +1,723 @@
...
+          return bigQueryRpc.deleteDataset(datasetId.dataset(), optionsMap);
+        }
+      }, options().retryParams(), EXCEPTION_HANDLER);
+    } catch (RetryHelper.RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public boolean delete(final String datasetId, final String tableId) throws BigQueryException {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: Nit: I believe the `final` is unnecessary here

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 25]
```diff
@@ -0,0 +1,982 @@
...
+  private static final BigQuery.DatasetDeleteOption DATASET_DELETE_CONTENTS =
+      BigQuery.DatasetDeleteOption.deleteContents();
+  private static final Map<BigQueryRpc.Option, ?> DATASET_DELETE_OPTIONS = ImmutableMap.of(
+      BigQueryRpc.Option.DELETE_CONTENTS, true);
+
+  // Table options
+  private static final BigQuery.TableOption TABLE_OPTION_FIELDS =
+      BigQuery.TableOption.fields(BigQuery.TableField.SCHEMA, BigQuery.TableField.ETAG);
+
+  // Table list options
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Could add the "startIndex" option here.

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 26]
```diff
@@ -0,0 +1,982 @@
...
+      BigQuery.TableDataListOption.startPageToken("cursor");
+  private static final Map<BigQueryRpc.Option, ?> TABLE_DATA_LIST_OPTIONS = ImmutableMap.of(
+      BigQueryRpc.Option.MAX_RESULTS, 42L,
+      BigQueryRpc.Option.PAGE_TOKEN, "cursor");
+
+  // Job options
+  private static final BigQuery.JobOption JOB_OPTION_FIELDS =
+      BigQuery.JobOption.fields(BigQuery.JobField.USER_EMAIL);
+
+  // Job list options
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Could test the "max results" and "start page token" options here too

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 27]
```diff
@@ -0,0 +1,982 @@
...
+      BigQuery.JobListOption.fields(BigQuery.JobField.STATISTICS);
+  private static final BigQuery.JobListOption JOB_LIST_ALL_USERS =
+      BigQuery.JobListOption.allUsers();
+  private static final BigQuery.JobListOption JOB_LIST_STATE_FILTER =
+      BigQuery.JobListOption.stateFilter(JobStatus.State.DONE, JobStatus.State.PENDING);
+  private static final Map<BigQueryRpc.Option, ?> JOB_LIST_OPTIONS = ImmutableMap.of(
+      BigQueryRpc.Option.ALL_USERS, true,
+      BigQueryRpc.Option.STATE_FILTER, ImmutableList.of("done", "pending"));
+
+  // Query Results options
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: If you think it's useful, we could add the max results and start page token options here

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 28]
```diff
@@ -0,0 +1,982 @@
...
+    String selector = (String) capturedOptions.getValue().get(TABLE_OPTION_FIELDS.rpcOption());
+    assertTrue(selector.contains("tableReference"));
+    assertTrue(selector.contains("schema"));
+    assertTrue(selector.contains("etag"));
+    assertEquals(31, selector.length());
+    assertEquals(updatedTableInfoWithProject, table);
+  }
+
+  @Test
+  public void testInsertAll() {
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Do you think it'd be worthwhile to test that `skipInvalidRows` and `ignoreUnknownValues` are being handled properly in this method?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 29]
```diff
@@ -0,0 +1,982 @@
...
+    assertTrue(bigquery.cancel(JobId.of(PROJECT, JOB)));
+  }
+
+  @Test
+  public void testQueryRequest() {
+    JobId queryJob = JobId.of(PROJECT, JOB);
+    com.google.api.services.bigquery.model.QueryResponse responsePb =
+        new com.google.api.services.bigquery.model.QueryResponse()
+            .setJobReference(queryJob.toPb())
+            .setJobComplete(false);
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Do you think we should also test the "true" case to check that the response is being setup correctly in that case?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java [hunk 30]
```diff
@@ -0,0 +1,982 @@
...
+            .setJobComplete(true)
+            .setCacheHit(false)
+            .setPageToken("cursor")
+            .setTotalBytesProcessed(42L)
+            .setTotalRows(BigInteger.valueOf(1L));
+    EasyMock.expect(bigqueryRpcMock.getQueryResults(JOB, EMPTY_RPC_OPTIONS)).andReturn(responsePb);
+    EasyMock.replay(bigqueryRpcMock);
+    bigquery = options.service();
+    QueryResponse response = bigquery.getQueryResults(queryJob);
+    assertEquals(queryJob, response.jobId());
```
Comment (ajkannan) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: Should we assert that `etag` is as expected as well?

File: gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java (hunk 25 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/test/java/com/google/gcloud/bigquery/BigQueryImplTest.java]: I guess you mean to add id to table data options.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 31]
```diff
@@ -25,5 +34,602 @@
...
+
+  /**
+   * Returns the requested job or {@code null} if not found.
+   *
+   * @throws BigQueryException upon failure
+   */
+  JobInfo getJob(JobId jobId, JobOption... options) throws BigQueryException;
+
+  /**
+   * Lists the dataset's tables.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Lists the jobs

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 32]
```diff
@@ -25,5 +34,602 @@
...
+   * @return {@code true} if cancel was requested successfully, {@code false} if the job was not
+   *     found
+   * @throws BigQueryException upon failure
+   */
+  boolean cancel(String jobId) throws BigQueryException;
+
+  /**
+   * Sends a job cancel request. This call will return immediately. The client will need to poll
+   * for the job status using either {@link #getJob(JobId, JobOption...)} or
+   * {@link #getJob(String, JobOption...)}) to see if the cancel operation completed successfully.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: what does it mean "completed **successfully**" for cancel?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java [hunk 33]
```diff
@@ -25,5 +34,602 @@
...
+
+  /**
+   * Runs the query associated to the request.
+   *
+   * @throws BigQueryException upon failure
+   */
+  QueryResponse query(QueryRequest request) throws BigQueryException;
+
+  /**
+   * Returns results of the query associated to the provided job.
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: s/ to / with / sounds better to me.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java [hunk 34]
```diff
@@ -0,0 +1,723 @@
...
+import com.google.gcloud.bigquery.InsertAllRequest.RowToInsert;
+import com.google.gcloud.spi.BigQueryRpc;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+final class BigQueryImpl extends BaseService<BigQueryOptions> implements BigQuery {
+
+  private static final Interceptor EXCEPTION_HANDLER_INTERCEPTOR = new Interceptor() {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: After your #438 we could probably handle this in a common way...

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java [hunk 35]
```diff
@@ -0,0 +1,723 @@
...
+    });
+    requestPb.setRows(rowsPb);
+    try {
+      return InsertAllResponse.fromPb(runWithRetries(new Callable<TableDataInsertAllResponse>() {
+        @Override
+        public TableDataInsertAllResponse call() {
+          return bigQueryRpc.insertAll(tableId.dataset(), tableId.table(), requestPb);
+        }
+      }, options().retryParams(), EXCEPTION_HANDLER));
+    } catch (RetryHelper.RetryHelperException e) {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: One problem with retries here is that we can get duplicates if `insertId` is missing.

Some options:
1) document that when retries are configured make sure you provide id to avoid duplicates.

2) We should add a `UUID.randomUUID().toString()` when `insertId` is missing (before the `runWithRetries`) call.

3) Don't use retry handler here.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java [hunk 36]
```diff
@@ -0,0 +1,723 @@
...
+            new Function<TableRow, List<FieldValue>>() {
+              @Override
+              public List<FieldValue> apply(TableRow rowPb) {
+                return Lists.transform(rowPb.getF(), FieldValue.FROM_PB_FUNCTION);
+              }
+            }));
+  }
+
+  @Override
+  public JobInfo getJob(final String jobId, JobOption... options) throws BigQueryException {
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: redundant `final`

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 11 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: I think it is useful for instance do avoid getting `statistics` if we do not need them. It becomes even more useful now that [queryPlan](https://cloud.google.com/bigquery/docs/reference/v2/jobs#statistics.query.queryPlan) is included in query statistics.

We do not provide a selector for list operation for the moment, but as soon as lists return complete resources I think we should add it.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java (hunk 35 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: We better go with 3) now, open an issue and consider 2)

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java [hunk 37]
```diff
@@ -0,0 +1,723 @@
...
+      return builder.build();
+    } catch (RetryHelper.RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+  }
+
+  @Override
+  public QueryResponse getQueryResults(JobId job, QueryResultsOption... options)
+      throws BigQueryException {
+    final Map<BigQueryRpc.Option, ?> optionsMap = optionMap(options);
```
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: no need for final.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 5 above)
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: unanswered.

**Comment by mziccard** [comment]
@aozarov I removed the remaining HashSet in place.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java (hunk 35 above)
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: Sounds good. did you create the issue?

**Comment by aozarov** [comment]
Looks good to me. @ajkannan do you have any more comments?

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 5 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: `code` is `400` and `message` is `Dataset gcloud-devel:dataset is still in use`. Once we tie `BigQueryError` to `BigQueryException` the exception will also have the `reason` information, which is `resourceInUse`.

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java (hunk 35 above)
Comment (mziccard) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQueryImpl.java]: Just opened #470

File: gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java (hunk 5 above)
Comment (aozarov) [inline gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/BigQuery.java]: Ok, it would be nice to include that in the Javadoc (though can wait until you tie BigQueryError to the exception).

**Comment by ajkannan** [comment]
Looks good to me as well
